File: MajorityElementMoreThanNDividedBy3/MajorityElementMoreThanNDividedBy3.cpp
/*
https://www.geeksforgeeks.org/problems/majority-vote/1
*/

class Solution {
  public:
    vector<int> findMajority(vector<int>& arr) {
        int sz=arr.size();
        map<int, int>mp;
        for(auto &vals: arr) mp[vals]++;
        vector<int>res;
        for(auto &entries: mp){
            if(entries.second>floor(sz/3)){
                res.push_back(entries.first);
            }
        }
        return res;
    }
};

File: MajorityElementMoreThanNDividedBy3/MajorityElementMoreThanNDividedBy3.java
/*
https://www.geeksforgeeks.org/problems/majority-vote/1
*/

class Solution {
    public ArrayList<Integer> findMajority(int[] arr) {
        int lengthOfArr=arr.length;
        HashMap<Integer, Integer>countMap=new HashMap<>();
        for(int i=0; i<lengthOfArr; ++i){
            if(countMap.containsKey(arr[i])){
                countMap.put(arr[i], countMap.get(arr[i])+1);
            } else countMap.put(arr[i], 1);
        }
        ArrayList<Integer>resultantList=new ArrayList<>();
        for(Map.Entry<Integer, Integer>countMapEntry: countMap.entrySet()){
            if(countMapEntry.getValue()>Math.floor(lengthOfArr)/3){
                resultantList.add(countMapEntry.getKey());
            }
        }
        Collections.sort(resultantList);
        return resultantList;
    }
}
